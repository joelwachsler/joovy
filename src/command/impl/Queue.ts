import { Message } from 'discord.js'
import { Environment } from '../../environment'
import { Command } from '../command'

export class Queue implements Command {
  argument = Command.ArgParser.create('queue')
  helpText = 'Print the current queue.'

  constructor(private env: Environment) {}

  handleMessage(_: Message) {
    this.env.printQueueRequest.next(null)
  }
}
