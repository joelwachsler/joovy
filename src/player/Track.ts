import JEvent from '../jevent/JEvent'
import { YtSearchResult } from '../jevent/YtSearchResult'

export default interface Track {
  name: string
  link: string
  removed: boolean
}

interface TrackArgs {
  event: JEvent
  info: YtSearchResult
}

export const from = ({ event, info }: TrackArgs): Track => {
  return {
    link: info.url,
    name: `[${info.title} (${info.timestamp})](${info.url}) [<@${event.message.author.id}>]`,
    removed: false,
  }
}
