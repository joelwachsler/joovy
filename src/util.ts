
export namespace Util {

  const apply = <T>(t: T) => {
  }
}
