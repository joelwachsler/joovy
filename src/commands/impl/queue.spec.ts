import Track from '../../player/Track'
import { playlistConfig } from '../../playlist/Playlist'
import { createTestEvent, handle, hot } from '../../test/util'
import { createReactions, FormatQueueArgs, QueueReactions } from './Queue'

test('should queue song if another one is already playing', () => {
  const play = createTestEvent({ content: '/play test' })
  const playAgain = createTestEvent({ content: '/play test2' })
  const playAgainAgain = createTestEvent({ content: '/play test3' })

  const messages = handle(hot('abc', { a: play, b: playAgain, c: playAgainAgain }))
  expect(messages).toMatchSnapshot()
})

test('should print queue with no tracks', () => {
  const queue = createTestEvent({ content: '/queue' })

  const messages = handle(hot('a', { a: queue }))
  expect(messages).toMatchSnapshot()
})

test('should print queue', () => {
  const play = createTestEvent({ content: '/play test' })
  const queue = createTestEvent({ content: '/queue' })

  const messages = handle(hot('ab', { a: play, b: queue }))
  expect(messages).toMatchSnapshot()
})

test('should print queue with multiple tracks', () => {
  const play = createTestEvent({ content: '/play test' })
  const playAnotherTrack = createTestEvent({ content: '/play test2' })
  const queue = createTestEvent({ content: '/queue' })

  const messages = handle(hot('abc', { a: play, b: playAnotherTrack, c: queue }))
  expect(messages).toMatchSnapshot()
})

test('should mark the current track the user is listening to', () => {
  const play = createTestEvent({ content: '/play test' })
  const playAnotherTrack = createTestEvent({ content: '/play test2' })
  const queue = createTestEvent({ content: '/queue' })

  // each track is 5 frames long, move to the frame after next track has started
  // i.e. frame 6 (ab = 2 frames + 4 frames = 6)
  const messages = handle(hot('ab----c', { a: play, b: playAnotherTrack, c: queue }))
  expect(messages).toMatchSnapshot()
})

test('should disconnect after the player has been idle for a certain amount of frames', () => {
  const play1 = createTestEvent({ content: '/play test1' })
  const play2 = createTestEvent({ content: '/play test2' })
  playlistConfig.timeoutFrames = 3

  // A song is 5 frame long, added on frame 0 -> frame 5 should result in the player being idle.
  // At frame 13 the player should disconnect.
  const messages = handle(hot('ab', { a: play1, b: play2 }))
  expect(messages).toMatchSnapshot()
})

test('should be able to reconnect after idle timeout', () => {
  const play = createTestEvent({ content: '/play test' })
  const skip = createTestEvent({ content: '/skip' })
  playlistConfig.timeoutFrames = 3

  const messages = handle(hot('ab----c', { a: play, b: skip, c: play }))
  expect(messages).toMatchSnapshot()
})

test('should add reactions to playlist longer than a single page', () => {
  const play1 = createTestEvent({ content: '/play test1' })
  const play2 = createTestEvent({ content: '/play test2' })
  const play3 = createTestEvent({ content: '/play test3' })
  const play4 = createTestEvent({ content: '/play test4' })
  const play5 = createTestEvent({ content: '/play test5' })
  const play6 = createTestEvent({ content: '/play test6' })
  const queue = createTestEvent({ content: '/queue' })

  const messages = handle(hot('abcdefg', {
    a: play1,
    b: play2,
    c: play3,
    d: play4,
    e: play5,
    f: play6,
    g: queue,
  }))
  expect(messages).toMatchSnapshot()
})

describe('reactions', () => {
  const createTrack = (input?: Partial<Track>): Track => {
    return {
      link: 'testLink',
      name: 'testName',
      removed: false,
      ...input,
    }
  }

  describe('with less than a single page', () => {
    it('should not print any reactions', () => {
      const args = new FormatQueueArgs([createTrack()], 0)
      const reactions = createReactions(args)

      expect(reactions).toEqual([])
    })
  })

  describe('with more than a single page', () => {
    // A single page has a size of 5, to get reactions we need more
    // than a single page, i.e. 6 tracks.
    const queue = Array(6).fill(createTrack())

    it('should react with next when on first page', () => {
      const reactions = createReactions(new FormatQueueArgs(queue, 0))
      expect(reactions).toEqual([QueueReactions.NEXT])
    })

    it('should react with previous when on seconds page', () => {
      const reactions = createReactions(new FormatQueueArgs(queue, 0).nextPage)
      expect(reactions).toEqual([QueueReactions.BACK])
    })
  })

  describe('with three pages', () => {
    // Three pages = 5 * 3 = 15
    const queue = Array(15).fill(createTrack())

    it('should react with next and two next when on first page', () => {
      const reactions = createReactions(new FormatQueueArgs(queue, 0))
      expect(reactions).toEqual([QueueReactions.NEXT, QueueReactions.TWO_NEXT])
    })

    it('should react with previous and next on second page', () => {
      const reactions = createReactions(new FormatQueueArgs(queue, 0).nextPage)
      expect(reactions).toEqual([QueueReactions.BACK, QueueReactions.NEXT])
    })

    it('should react with two previous and previous on third page', () => {
      const reactions = createReactions(new FormatQueueArgs(queue, 0).nextPage.nextPage)
      expect(reactions).toEqual([QueueReactions.TWO_BACK, QueueReactions.BACK])
    })
  })
})