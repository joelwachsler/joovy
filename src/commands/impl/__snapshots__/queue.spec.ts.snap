// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add reactions to playlist longer than a single page 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test1 (1:10)](https://test1.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test1 (1:10)](https://test1.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test3 (1:10)](https://test3.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test4 (1:10)](https://test4.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test5 (1:10)](https://test5.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test6 (1:10)](https://test6.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 6,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 6,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"Queue\\",
  \\"type\\": \\"rich\\",
  \\"description\\": \\"\`0)\` [test1 (1:10)](https://test1.com) [<@testAuthorId>]\\\\n\`1)\` [test2 (1:10)](https://test2.com) [<@testAuthorId>] ← Currently playing\\\\n\`2)\` [test3 (1:10)](https://test3.com) [<@testAuthorId>]\\\\n\`3)\` [test4 (1:10)](https://test4.com) [<@testAuthorId>]\\\\n\`4)\` [test5 (1:10)](https://test5.com) [<@testAuthorId>]\\\\n\`5)\` [test6 (1:10)](https://test6.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": \\"2000-02-01T00:00:00.000Z\\",
  \\"color\\": 16758815,
  \\"fields\\": [
    {
      \\"name\\": \\"\\\\n1 more track(s) in queue.\\",
      \\"value\\": \\"During the current session 6 tracks have been added to the queue in total\\",
      \\"inline\\": false
    }
  ],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 6,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "addedReactions": Array [
          "▶",
        ],
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test3 (1:10)](https://test3.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 15,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 15,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test4 (1:10)](https://test4.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 20,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 20,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test5 (1:10)](https://test5.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 25,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 25,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test6 (1:10)](https://test6.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 30,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`should mark the current track the user is listening to 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 6,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 6,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"Queue\\",
  \\"type\\": \\"rich\\",
  \\"description\\": \\"\`0)\` [test (1:10)](https://test.com) [<@testAuthorId>]\\\\n\`1)\` [test2 (1:10)](https://test2.com) [<@testAuthorId>] ← Currently playing\\",
  \\"url\\": null,
  \\"timestamp\\": \\"2000-02-01T00:00:00.000Z\\",
  \\"color\\": 16758815,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`should print queue 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"Queue\\",
  \\"type\\": \\"rich\\",
  \\"description\\": \\"\`0)\` [test (1:10)](https://test.com) [<@testAuthorId>] ← Currently playing\\",
  \\"url\\": null,
  \\"timestamp\\": \\"2000-02-01T00:00:00.000Z\\",
  \\"color\\": 16758815,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`should print queue with multiple tracks 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"Queue\\",
  \\"type\\": \\"rich\\",
  \\"description\\": \\"\`0)\` [test (1:10)](https://test.com) [<@testAuthorId>] ← Currently playing\\\\n\`1)\` [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": \\"2000-02-01T00:00:00.000Z\\",
  \\"color\\": 16758815,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`should print queue with no tracks 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"The queue is empty 👀\\",
  \\"type\\": \\"rich\\",
  \\"description\\": null,
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should queue song if another one is already playing 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test3 (1:10)](https://test3.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test3 (1:10)](https://test3.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 15,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;
